milli/src/search/query_tree.rs

```diff
@@ -580,17 +611,24 @@ fn create_matching_words(
                     TypoConfig { max_typos: 2, word_len_one_typo, word_len_two_typo, exact_words };
 
                 let matching_word = match typos(word, authorize_typos, config) {
-                    QueryKind::Exact { word, .. } => MatchingWord::new(word, 0, prefix),
-                    QueryKind::Tolerant { typo, word } => MatchingWord::new(word, typo, prefix),
+                    QueryKind::Exact { word, .. } => matching_word_cache.insert(word, 0, prefix),
+                    QueryKind::Tolerant { typo, word } => {
+                        matching_word_cache.insert(word, typo, prefix)
+                    }
                 };
-                matching_words.push((vec![matching_word], vec![id]));
+                if let Some(matching_word) = matching_word {
+                    matching_words.push((vec![matching_word], vec![id]));
+                }
             }
             // create a CONSECUTIVE matchings words wrapping all word in the phrase
             PrimitiveQueryPart::Phrase(words) => {
                 let ids: Vec<_> =
                     (0..words.len()).into_iter().map(|i| id + i as PrimitiveWordId).collect();
-                let words =
-                    words.into_iter().flatten().map(|w| MatchingWord::new(w, 0, false)).collect();
+                let words = words
+                    .into_iter()
+                    .flatten()
+                    .flat_map(|w| matching_word_cache.insert(w, 0, false))
+                    .collect();
                 matching_words.push((words, ids));
```

Comment:
Same comment as for `synonym` on line 593.

Reply:
so I'll do the same Answer !